File: 46.全排列.swift
/*
 * @lc app=leetcode.cn id=46 lang=swift
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        var result = [[Int]]()
        func permute(prefix: [Int], remain: [Int]){
            guard remain.count > 1 else {
                var sum = prefix
                sum.append(contentsOf: remain)
                result.append(sum)
                return
            }
            
            for i in 0..<remain.count {
                var nextRemain = remain
                nextRemain.remove(at: i)
            
                var nextPrefix = prefix
                nextPrefix.append(remain[i])
                permute(prefix: nextPrefix, remain: nextRemain)
            }
        }
        
        permute(prefix: [], remain: nums)
        
        return result
    }
}
// @lc code=end

File: 46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.helper([], nums, result)
        return result
        
    def helper(self, prefix: List[int], remain: List[int], result: List[List[int]]):
        if len(remain) == 1:
            prefix.extend(remain)
            result.append(prefix)
            return 
        
        for value in remain:
            prefixCopy = prefix.copy()
            prefixCopy.append(value)
            remainCopy = remain.copy()
            remainCopy.remove(value)
            self.helper(prefixCopy, remainCopy, result)
            
# @lc code=end

